# Imports

In [1]:
import pandas as pd
import numpy as np 

---

# Events - Carga de datos

In [2]:
events = pd.read_csv("../../data/tp2/events.csv")

/opt/anaconda/3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (13,20) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


---

## Inspección de columnas y tipos

In [3]:
events.dtypes

index                   int64
date                   object
event_id                int64
ref_type                int64
ref_hash                int64
application_id          int64
attributed               bool
device_countrycode      int64
device_os_version     float64
device_brand          float64
device_model          float64
device_city           float64
session_user_agent    float64
trans_id               object
user_agent            float64
event_uuid             object
carrier               float64
kind                  float64
device_os             float64
wifi                     bool
connection_type        object
ip_address              int64
device_language       float64
dtype: object

In [5]:
len(events)

7744581

In [4]:
events.memory_usage(deep=True).sum() / (1024**2)

2829.647403717041

In [6]:
exploracion = events.isna().any()
exploracion = exploracion.to_frame()

In [7]:
exploracion.columns = ["tiene_nulos"]

In [8]:
exploracion["todos_nulos"] = events.isna().all()

In [12]:
exploracion["cantidad_nulos"] = len(events) - events.count()

In [15]:
exploracion["porcentaje_nulos"] = exploracion["cantidad_nulos"] / len(events)

In [21]:
exploracion["porcentaje_nulos"] = exploracion["porcentaje_nulos"].map(lambda x: round(x*100, 2))

In [30]:
exploracion["maximo"] = events.max()

In [31]:
exploracion["minimo"] = events.min()

In [34]:
exploracion["dtype"] = events.dtypes

In [40]:
exploracion["valores_unicos"] = events.nunique()

In [44]:
exploracion["porcentaje_valores_unicos"] = (exploracion["valores_unicos"] / len(events)) *100 
exploracion["porcentaje_valores_unicos"] = exploracion["porcentaje_valores_unicos"].map(lambda x: round(x, 2))

In [55]:
exploracion

tiene_nulos  todos_nulos  cantidad_nulos  \
index                     False        False               0   
date                      False        False               0   
event_id                  False        False               0   
ref_type                  False        False               0   
ref_hash                  False        False               0   
application_id            False        False               0   
attributed                False        False               0   
device_countrycode        False        False               0   
device_os_version          True        False         5411606   
device_brand               True        False         5191157   
device_model               True        False         2076489   
device_city                True        False         5849646   
session_user_agent         True        False           42280   
trans_id                   True        False         7706939   
user_agent                 True        False         4403098   
event_uuid                 True        False           29772   
carrier                    True        False         5818680   
kind                       True        False           29772   
device_os                  True        False         5874391   
wifi                      False        False               0   
connection_type            True        False         5935285   
ip_address                False        False               0   
device_language            True        False         2079172   

                    porcentaje_nulos                   maximo  \
index                           0.00                  2619098   
date                            0.00  2019-04-26 23:59:59.881   
event_id                        0.00                      576   
ref_type                        0.00      1891515180541284343   
ref_hash                        0.00      9223365873867066370   
application_id                  0.00                      370   
attributed                      0.00                     True   
device_countrycode              0.00      6287817205707153877   
device_os_version              69.88              9.13548e+18   
device_brand                   67.03              9.22151e+18   
device_model                   26.81              9.22336e+18   
device_city                    75.53              9.22137e+18   
session_user_agent              0.55              9.22142e+18   
trans_id                       99.51                      NaN   
user_agent                     56.85              9.22276e+18   
event_uuid                      0.38                      NaN   
carrier                        75.13              9.22046e+18   
kind                            0.38              9.22002e+18   
device_os                      75.85              7.53167e+18   
wifi                            0.00                     True   
connection_type                76.64                      NaN   
ip_address                      0.00      9223349765218515215   
device_language                26.85              9.21023e+18   

                                     minimo    dtype  valores_unicos  \
index                                     0    int64         1567735   
date                2019-04-18 00:00:00.027   object         7693028   
event_id                                  0    int64             516   
ref_type                1494519392962156891    int64               2   
ref_hash                     40621409780134    int64          362619   
application_id                            1    int64             250   
attributed                            False     bool               2   
device_countrycode      6287817205707153877    int64               1   
device_os_version               1.00408e+16  float64              96   
device_brand                    4.35915e+15  float64             506   
device_model                    7.96735e+14  float64            4720   
device_city                     6.48007e+15  float64      

***

## Conclusiones para events:
- __index__: cargar como int32. No hay nulos ni anómalos.
- __date__: cargar como datetime64. No hay nulos ni anómalos.
- __event_id__: cargar como int16, no hay valores nulos.
- __ref_type__: No tiene nulos y hay solo dos valores posibles. Debería convertirse el más frecuente a google y el otro a apple.
- __ref_hash__: cargar como int64, no hay valores nulos.
- __application_id__: cargar como int16, no hay valores nulos.
- __attributed__: cargar como bool, no hay valores nulos.
- __device_countrycode__: No cargar. Es todo uruguay y no hay nulos.
- __device_os_version__: Tiene el 70% de valores nulos, hay que ver qué tan significativa resulta. Si no aparece en otros datasets, se podría recodificar ya que hay solo 96 valores únicos.
- __device_brand__: idem __device_os_version__.
- __device_model__: idem __device_os_version__.
- __device_city__: 75% de nulos, habría que ver cuánto influye. si no está en otros datasets podría recodificarse para que no ocupe tanto.
- __session_user_agent__: Casi no hay valores nulos. No se podría tocar ya que se comparte con installs.
- __device_countrycode__: No cargar. Es todo uruguay y no hay nulos.
- __event_uuid__: Podría no cargarse, ya que el 99,62% de los valores son diferentes y además actúa como un índice.
- __carrier__: 75,13% de nulos, habría que ver si aporta. Parece raro que haya 259 carriers diferentes.
- __kind__: Casi no tiene nulos. Hay 513 tipos de evento diferentes.
- __wifi__: Cargarla como bool, no tiene nulos.
- __connection_type__: Tiene 4 valores posibles y 76% son nulos. Habría que ver si aporta.
- __ip_address__: No hay valores nulos, cargarla como int64 o reconvertir ha que hay 381.000 valores diferentes.
- __device_language__: Ver si aporta, tiene 209 valores posibles. Parece un número exagerado.


---